An integration to control relays over HTTP

1e1 opened the pull request:
(on second thought after https://github.com/home-assistant/core/pull/39343)
It's a NO-REST component to control switches. 
It sends commands by using path only:
  - so the embed webserver parses the first header only... yes, it's very light)
  - any command can be played by a browser


Add the following lines into `configuration.yaml`:
```
switch:
  - platform: http_inline
    host: http://webrelay.local
    relays:
      0:  relayname 1
      1:  relay name 2
      2:  a relay name 3
      30: another relayname 31
      31: and this relayname 32
```

Or use the integration wizard:

1- "user"
  - host
  - number of relays
  - name
  - READ path pattern
  - WRITE path pattern

2- "relay"
  - name 0
  - name 1
  - ...
  - name i

---

You can install an Arduino client from: https://github.com/1e1/arduino-webcontroller-relay

Control a device understanding routes:
- READ:  {host}/r/{relay_id}
- WRITE: {host}/w/{relay_id}/{state:[0-1]}

the response payload is:
"{state} {relay_id} {NC or NO} {pin_id}"

specs: https://raw.githubusercontent.com/1e1/arduino-webcontroller-relay/master/doc/swagger.yml

---

<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Breaking change

none


## Proposed change

1 new component (cf on top)


## Type of change

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Example entry for `configuration.yaml`:
<!--
  Supplying a configuration snippet, makes it easier for a maintainer to test
  your PR. Furthermore, for new integrations, it gives an impression of how
  the configuration would look like.
  Note: Remove this section if this PR does not have an example entry.
-->

```yaml
switch:
  - platform: http_inline
    host: http://webrelay.local
    relays:
      0:  relayname 1
      1:  relay name 2
      2:  a relay name 3
      30: another relayname 31
      31: and this relayname 32
```

## Additional information

- first draft aborted after discussion https://github.com/home-assistant/core/pull/39343
- swagger of the IoT client: https://raw.githubusercontent.com/1e1/arduino-webcontroller-relay/master/doc/swagger.yml

## Checklist

- [x] The code change is tested and works locally.
- [ ] Local tests pass. **Your PR cannot be merged unless tests pass**
- [ ] There is no commented out code in this PR.
- [ ] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [x] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [x] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [ ] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-asc+-review%3Aapproved

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

1e1 committed:
An integration to control relays over HTTP.
You can install an Arduino client from: https://github.com/1e1/arduino-webcontroller-relay

Control a device understanding routes:
- READ:  {host}/r/{relay_id}
- WRITE: {host}/w/{relay_id}/{state:[0-1]}

the response payload is:
"{state} {relay_id} {NC or NO} {pin_id}"

specs: https://raw.githubusercontent.com/1e1/arduino-webcontroller-relay/master/doc/swagger.yml

Add the following lines into `configuration.yaml`:
```
switch:
  - platform: http_plaintext
    host: http://webrelay.local
    relays:
      0:  relayname 1
      1:  relay name 2
      2:  a relay name 3
      30: another relayname 31
      31: and this relayname 32
```

homeassistant commented:
Hi @1e1, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;core#39558).

Once you do that we will be able to review and accept this pull request.

Thanks!

davet2001 commented:
Probably one of the core devs should comment here, but I’m not sure this type of integration is allowed.

The proposed integration seems to define the API within within home assistant core itself.  My understanding from 

https://developers.home-assistant.io/docs/creating_component_code_review#4-communication-with-devicesservices

...was that the actual interaction with end devices should be done by a 3rd party PyPi library.  I assume this is for maintenance reasons so that improvements/changes/testing can be done on that library independently rather than HA itself, then followed by a version bump in HA.

There are many many services that could be defined by one-liner http calls and they can be implemented without HA core code mods e.g. using 
https://www.home-assistant.io/integrations/switch.command_line
or
https://www.home-assistant.io/integrations/switch.rest/

I realise that what you’ve proposed is extremely lightweight, but that seems to be where the line is drawn

As I said, the decision is not up to me. Happy to be corrected.

frenck commented:
I think this integration is a duplicate, as our rest integration allows for just this.

I've seen a comment in the previous PR of this being simpler to configure, however, I do not agree on this. It is simple to configure because of a very limited and specific use-case. So now, instead of relays (which are called switches in the Home Assistant world btw), I would love to configure a light? with dimming capabilities? a binary sensor? climate?

The reason why it is simple is that it is limited to your use-case, however, for a wider adoption or more generic use-case, more is needed. While this might be a good start, extending it like that, would end up being a rest sensor again.

balloob commented:
I agree with @frenck. We should instead upgrade the rest switch to be able to use `GET` and use a resource template.

1e1 commented:
Hello All, 

Yes, I'm focus on my own use case. And it's how I found HA ;) 
I have started on Relay, I have to rename it for the HA context. 

I have already read AREST. For me, it's different. 
Conversely I didn't know COMMAND_LINE: it is very powerful! 

In both cases there is not wizard to setup a new entry. 

@frenck there is a cursor between simplicity and universal controller. 
I didn't meet HA users and I don't know the guiding principle. 

I will be happy to upgrade another component if it's better for the community. From my point of view the config_flow is very important.

What do you think?

frenck commented:
I think we should upgrade the rest component to allow for GET requests in switches. As these integrations define their protocol, a configuration UI is not required.

davet2001 commented:
I think @1e1 was hopeful about a UI/config_flow to configure this and I agree.

Having a config flow makes a component accessible to those not wanting to edit yaml or restart HA.

Is the solution an enhanced REST component with a new GET option that can also be configured via a config flow?

1e1 commented:
Hi @frenck @davet2001 

I'm not sure modifying the REST component is the best way. 
Because I target a no-REST server: the action is not is the HTTP method, it's the prefix of the path:
- 'r' for Read: /r/{relayId}
- 'w' for Write: /w/{relayId}/{state}

I can use command_line component: 
```yaml
switch:
  - platform: command_line
    switches:
      r0:
        command_on: "/usr/bin/curl -X GET http://webrelay.local/w/0/1"
        command_off: "/usr/bin/curl -X GET http://webrelay.local/w/0/0"
        command_state: "/usr/bin/curl -X GET http://webrelay.local/r/0"
        value_template: '{{ value.split()[0] == "1" }}'
        friendly_name: "Relay #0"
      r42:
        command_on: "/usr/bin/curl -X GET http://webrelay.local/w/42/1"
        command_off: "/usr/bin/curl -X GET http://webrelay.local/w/42/0"
        command_state: "/usr/bin/curl -X GET http://webrelay.local/r/42"
        value_template: '{{ value.split()[0] == "1" }}'
        friendly_name: "Relay #42"
```

In this case an UI for command_line should be better... 
On the other hand configuring a command line is risky for security, especially if it's easy to access.

frenck commented:
This is all very off-topic.

Looking at the integration contributed here, I think it is not build to scale/be flexible and mimics a large part of the REST integration, which can be adjusted to do just this.

This discussion is not about the UI at this point. Anything can have a config flow, however, we have a list of integrations that require that. This type of communications and the definition that is needed, do not require a config flow. Nothing forbids it, however, it is out of scope for this PR.

The security issue raised in your last comment is a false sense of security. If one has that level of access to HA, one can already access your whole system. A UI configuration for the command line isn't going to change that. However, out of scope IMHO.

The question is: This integration provided in this PR. We should not merge this or add this, based on the comments I initially gave. At this point, I stand by my comment to adjust the REST integration to allow for this. This change should be fairly minimal and provides greater flexibility.

1e1 commented:
@frenck 
I'm not sure a simple modification of the rest component could solve this case. 
The URL for reading is not the same than the one for writing.
I think the modification of the REST component will result in URLs looking like `/path/{sensor_id}?param_1=value_1&param_i=value_i`

In this case I would like have a *parameter replacement into the path* and *default value*. 
Example: 
resource: `http://webrelay.local/{action='r'}/42/{state}`
body_on: `{"action": "w", "state": 1, "param_i": "value_i"}`
body_off: `{"action": "w", "state": 0, "param_i": "value_i"}`

should generate: 
state_url: `http://webrelay.local/r/42/`  # default value for "action" and null for other parameters
on_url: `http://webrelay.local/w/42/1?action=w&state=1&param_i=value_i`  # or parameters already replaced can be remove from the query string ;) 
off_url: `http://webrelay.local/w/42/0?action=w&state=1&param_i=value_i`

These features are more important than the `GET` method; the webrelay server skip the word. 

How do you imagine the modification?

frenck commented:
Yes, it will need query parameters, which is being implemented in #38482 (and is lacking tests at this point).
